DAOS-15628-test: Verify maximum containers create with and without dup metadata ops

dinghwah opened the pull request:
Description: implement more test to metadata with disable svc.
 
Skip-unit-tests: true
Skip-fault-injection-test: true
Test-tag: ObjectMetadata
Doc-only: false
Required-githooks: true
Signed-off-by: Ding Ho ding-hwa.ho@intel.com

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-14145/1/testReport/

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-14145/5/testReport/

daosbuild1 commented:
Test stage Test RPMs on EL 8.6 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14145/6/display/redirect

kccain commented:
Jenkins build 8 https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14145/8/artifact/Functional%20Hardware%20Large/server/metadata.py/results.json seems to have run 2 configurations for test_metadata_addremove and test_metadata_server_restart, which is not desired. The total runtime went to 3500+ seconds, compared to a typical weekly run duration of 1500+ seconds (for example https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/378/artifact/Functional%20Hardware%20Large/server/metadata.py/results.json) . Only muxing for test_metadata_fillup is desired.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14145/10/execution/node/880/log

daltonbohning commented on src/tests/ftest/server/metadata.py:
```diff
@@ -343,9 +387,12 @@ def test_metadata_addremove(self):
-        if not self.run_dummy_metadata_workload():
-            self.fail("failed to run dummy metadata workload")
-
+        # Skip dummy_metadata_workload when feature is disabled
+        svc_ops_enabled = self.pool.get_property("svc_ops_enabled")
+        if svc_ops_enabled:
+            svc_ops_entry_age = self.pool.get_property("svc_ops_entry_age")
+            if not self.run_dummy_metadata_workload(duration=svc_ops_entry_age):
+                self.fail("failed to run dummy metadata workload")
```
We shouldn't conditionally run the test, and the default is enabled, so we should be able to just run this
```suggestion
        svc_ops_entry_age = self.pool.get_property("svc_ops_entry_age")
        if not self.run_dummy_metadata_workload(duration=svc_ops_entry_age):
            self.fail("failed to run dummy metadata workload")
```

dinghwah replied:
The suggestion does not work, since svc_ops_entry_age could be a valid number (ie: 60, or 300...) and svc_ops_enabled is disabled and causing issue.

daltonbohning replied:
This calls `self.create_pool()`
Which has
```
def create_pool(self, svc_ops_enabled=True):
...
    if svc_ops_enabled:
        self.add_pool()   
```
So `svc_ops_enabled` is turned on

dinghwah replied:
no.  With the suggested 390-392, when svc_ops_enabled=False, the run_metsdata_dummy_workload will be executed.

daltonbohning replied:
How would `svc_ops_enabled=False`? The default is True and this test doesn't pass `False`

dinghwah replied:
The testcase svc_ops_disabled will set svc_ops_enabled=False?

daltonbohning replied:
This is `test_metadata_addremove`

dinghwah replied:
Yes. To determine running dummy_workload should check for svc_ops_enabled or not, not svc_ops_entry_age.

daltonbohning replied:
`svc_ops_enabled` isn't dynamic here. It's always what is set in the config
https://github.com/daos-stack/daos/blob/05474cfc94b4018436bc648d6e3e442b2011d0f0/src/tests/ftest/server/metadata.yaml#L53

kccain replied:
For test_metadata_addremove(), for simplicity, I think we always want to run one variant (in this case, let's say we decide to use svc_ops_enabled:1 and run a dummy workload loop of svc_ops_entry_age seconds).

Which makes me wonder if in the .yaml file, maybe it should not specify svc_ops_enabled in the pool properties at all. And just leave it to the test cases in this code to specify in the argument to add_pool() whether svc_ops_enabled shall be True or False?

test_metadata_fillup_svc_ops_disabled and test_metadata_fillup_svc_ops_enabled already pass in the argument explicity.
test_metadata_addremove already effectively does the same (no argument means the default svc_ops_enabled=True is used).

daltonbohning commented on src/tests/ftest/server/metadata.py:
```diff
@@ -86,9 +86,19 @@ def pre_tear_down(self):
+        """Create a pool and display the svc ranks.
+
+        Args:
+            svc_ops_enabled (Bool, optional): pool create with svc_ops_enabled, default to True.
```
```suggestion
            svc_ops_enabled (bool, optional): pool create with svc_ops_enabled. Defaults to True.
```

daltonbohning commented on src/tests/ftest/server/metadata.py:
```diff
@@ -86,9 +86,19 @@ def pre_tear_down(self):
+            svc_ops_enabled (Bool, optional): pool create with svc_ops_enabled, default to True.
+
+        """
+        if svc_ops_enabled:
+            self.add_pool()
+        else:
+            params = {}
+            params['properties'] = "svc_ops_enabled:0"
+            self.add_pool(**params)
```
FYI
```suggestion
        params = {}
        if not svc_ops_enabled:
            params['properties'] = "svc_ops_enabled:0"
        self.add_pool(**params)
```

daltonbohning commented on src/tests/ftest/server/metadata.py:
```diff
@@ -236,24 +246,21 @@ def run_dummy_metadata_workload(self, duration=150):
-        Use Cases:
-            ?
+        Args:
+            svc_ops_enabled (bool): Pool create properties svc_ops_enabled, default to True.
```
```suggestion
            svc_ops_enabled (bool): Pool create properties svc_ops_enabled. Defaults to True.
```

dinghwah replied:
Updated.

daltonbohning commented on src/tests/ftest/server/metadata.py:
```diff
@@ -236,24 +246,21 @@ def run_dummy_metadata_workload(self, duration=150):
-        svc_ops_entry_age = self.pool.get_property("svc_ops_entry_age")
-        if not self.run_dummy_metadata_workload(duration=svc_ops_entry_age):
-            self.fail("failed to run dummy metadata workload")
+        self.log_step("1. Create pool with properties svc_ops_enabled: {}".format(svc_ops_enabled))
```
We shouldn't be adding numbers when using `log_step` because `log_step` already adds numbers

dinghwah replied:
The reason to add 1. 2. 3.... for the step is sync with the test steps in the testcase description.

daltonbohning replied:
But this output is confusing
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14145/12/artifact/Functional%20Hardware%20Large/server/metadata.py/steps.log
```
Step 5: 1. Create pool with properties svc_ops_enabled: True [elapsed since last step: 0.00s]
Step 6: 2. sustained container creates: to no space and beyond. [elapsed since last step: 66.16s]
Step 7: 3. Verify number of container within limit. [elapsed since last step: 114.49s]
Step 8: 4. Cleaning up 500 containers after pool is full. [elapsed since last step: 0.00s]
```

dinghwah replied:
I think it's clear. the global Step 4, and the testcase step 1.

daltonbohning replied:
@phender opinion?

daltonbohning commented on src/tests/ftest/server/metadata.py:
```diff
@@ -236,24 +246,21 @@ def run_dummy_metadata_workload(self, duration=150):
+            self.log.info("svc_ops_enabled enabled, run dummy_metadata_workload")
+            svc_ops_entry_age = self.pool.get_property("svc_ops_entry_age")
+            if not self.run_dummy_metadata_workload(duration=svc_ops_entry_age):
+                self.fail("#1. failed to run dummy metadata workload")
```
Similar - we shouldn't add numbers to `self.fail` because they probably won't correspond to the numbers that `log_step` automatically adds

dinghwah replied:
Same.

daltonbohning commented on src/tests/ftest/server/metadata.yaml:
```diff
@@ -47,7 +48,16 @@ pool:
   scm_size: 1G
   control_method: dmg
-  properties: svc_ops_enabled:1,svc_ops_entry_age:150
+  pool_prop: !mux
+    svc_ops_60:
+      properties: svc_ops_enabled:1,svc_ops_entry_age:60
```
We should simplify this since we only have 1 variant running
```suggestion
  properties: svc_ops_enabled:1,svc_ops_entry_age:60
```

dinghwah replied:
@kccain, please comment.

kccain commented on src/tests/ftest/server/metadata.yaml:
```diff
@@ -47,7 +48,16 @@ pool:
+  pool_prop: !mux
+    svc_ops_60:
+      properties: svc_ops_enabled:1,svc_ops_entry_age:60
+# Uncomment following for manual test
```
Also might add that the values of created_cont_min and created_cont_max will be affected by the duration of the svc_ops_entry_age, so they may need to be adjusted accordingly.

kccain reviewed (commented):
added my comments, I hope they are helpful. There is a lot of activity in the PR

I do realize this one was closed, not sure why. But I figured it would be more valuable to add my comments here with the other comments submitted so far.

dinghwah commented:
Thanks for the comments. 
After discussed with Phil and met with Ken:
https://github.com/daos-stack/daos/pull/14243 should have addressed all the above issues and comments. Please review PR-14243 if I missed anything.
